test/credentials.spec.js

```diff
@@ -580,6 +580,56 @@
           return done();
         });
       });
+      it('default tokenCodeFn works if mfa_serial is provided', function(done) {
+        var creds, tokenCodeFnSpy, mock;
+        mock = '[default]\naws_access_key_id = akid\naws_secret_access_key = secret\naws_session_token = session\n'
+          + '[profile withmfa]\nrole_arn = arn\nmfa_serial = serial\nsource_profile = default';
+        helpers.spyOn(AWS.util, 'readFileSync').andReturn(mock);
+        helpers.mockHttpResponse(200, {}, '<AssumeRoleResponse xmlns="https://sts.amazonaws.com/doc/2011-06-15/">\n  <AssumeRoleResult>\n    <Credentials>\n      <AccessKeyId>KEY</AccessKeyId>\n      <SecretAccessKey>SECRET</SecretAccessKey>\n      <SessionToken>TOKEN</SessionToken>\n      <Expiration>1970-01-01T00:00:00.000Z</Expiration>\n    </Credentials>\n  </AssumeRoleResult>\n</AssumeRoleResponse>');
+        creds = new AWS.SharedIniFileCredentials({ profile: 'withmfa' });
+        tokenCodeFnSpy = helpers.createSpy('tokenCodeFn');
+        creds.refresh(function(e) {
+          done();
+        });
+        expect(tokenCodeFnSpy.calls.length).to.equal(1);
+      });
+      it('calls tokenCodeFn if mfa_serial is provided', function(done) {
+        var creds, tokenCodeFn, mock, tokenCodeFnSpy;
+        mock = '[default]\nrole_arn = arn\nmfa_serial = serial\naws_access_key_id = key\naws_secret_access_key = secret\n'
+          + '[profile withmfa]\nrole_arn = arn\nmfa_serial = serial\nsource_profile = default';
+        helpers.spyOn(AWS.util, 'readFileSync').andReturn(mock);
+        helpers.mockHttpResponse(200, {}, '<AssumeRoleResponse xmlns="https://sts.amazonaws.com/doc/2011-06-15/">\n  <AssumeRoleResult>\n    <Credentials>\n      <AccessKeyId>KEY</AccessKeyId>\n      <SecretAccessKey>SECRET</SecretAccessKey>\n      <SessionToken>TOKEN</SessionToken>\n      <Expiration>1970-01-01T00:00:00.000Z</Expiration>\n    </Credentials>\n  </AssumeRoleResult>\n</AssumeRoleResponse>');
+        tokenCodeFnSpy = helpers.createSpy('tokenCodeFn');
+        tokenCodeFn = function(serial, callback) {
+          tokenCodeFnSpy(serial);
+          callback(null, '123456');
+          done();
+        };
+        creds = new AWS.SharedIniFileCredentials({
+          profile: 'withmfa',
+          tokenCodeFn: tokenCodeFn
+        });
+        expect(tokenCodeFnSpy.calls.length).to.equal(1);
```

Comment:
Can you move these expect statements to inside of your `tokenCodeFn`, before `done()`? Right now `done()` is getting called before these expect statements do, so this test will never report a failure.